src/visit.rs

```diff
@@ -0,0 +1,211 @@
+use std::collections::BTreeMap;
+
+use crate::{FeatureValue, Metadata, Package};
+
+/// A visitor over a package's features and their dependencies.
+pub trait FeatureVisitor {
+    /// The error type of a walk.
+    type Error;
+
+    /// Visits a missing dependency.
+    ///
+    /// Return `Ok(())` to continue the walk, or `Err(…)` to abort it.
+    fn visit_missing_dependency(&mut self, dep_name: &str) -> Result<(), Self::Error>;
+
+    /// Visits a missing package.
+    ///
+    /// Return `Ok(())` to continue the walk, or `Err(…)` to abort it.
+    fn visit_missing_package(&mut self, pkg_name: &str) -> Result<(), Self::Error>;
+
+    /// Visits a feature that's enabling another feature.
+    ///
+    /// Return `Ok(<walk>)` where `<walk>` indicates whether or not to walk the feature's downstream dependencies,
+    /// or `Err(…)` to abort the walk.
+    fn visit_feature(
+        &mut self,
+        package: &Package,
+        feature_name: &str,
+    ) -> Result<bool, Self::Error> {
+        let (..) = (package, feature_name);
+        Ok(true)
+    }
+
+    /// Visits a feature that's enabling a dependency with `dep:dep_name` syntax.
+    ///
+    /// Return `Ok(<walk>)` where `<walk>` indicates whether or not to walk the feature's downstream dependencies,
+    /// or `Err(…)` to abort the walk.
+    fn visit_dep(&mut self, package: &Package, dep_name: &str) -> Result<bool, Self::Error> {
+        let (..) = (package, dep_name);
+        Ok(true)
+    }
+
+    /// Visits a feature that's enabling a feature on a dependency with `crate_name/feat_name` syntax.
```

Comment:
Both of these doc comments say "visits a feature ..." but they both seem to take `package: &Package` and nothing resembling a feature. I'm confused about how this interface works =/

Reply:
The three visitor methods correspond to the variants of the `FeatureValue` type (which was copied from `cargo` itself). Every time the walker passes such a value in the feature graph it calls the corresponding method on the visitor that it carries along.

The variants of said type unfortunately have rather confusing names already: `Feature`, `Dep`, `DepFeature`, especially when considering that `FeatureValue` is itself the value of a "feature dependency" (i.e. a feature depended on by another feature). Finding proper names for these was … a challenge. One that I decided to mostly gloss over for now. #bike-shedding.

Preferably the `FeatureValue` enum would have new-type variants, rather than struct variants, so that each variant's type could be passed to the corresponding visitor method and also so that the corresponding type of feature value could be handled as a single value. But! By promoting each variant to individual types each of them would have to have their own `Serialize`/`Deserialize`/`FromStr` implementations. An explosion of complexity and change that I wanted to avoid if possible. That said I do think that the end goal should be a new-type based enum.